### PR TITLE
[wip] Adds possessable var to living

### PR DESCRIPTION
**Pull request by JJRcop**
:cl: JJRcop
add: Objects that ghosts can interact with now have a purple particle effect.
/:cl:

This PR adds a new ondemand_possessable var. This allows ghosts to ctrl+click the mob, which lets them take over. Also adds particle effects for atoms ghosts can meaningfully interact with.

This was inspired by looking at #31630's diff, and the related [```replace_banned_player()```](https://github.com/tgstation/tgstation/pull/31630/files#diff-aaad5985ed9decf25a96c4935bde4269R69) refactor. I thought about this and possible similar situations where maybe you don't get a candidate within the time limit, but you still want someone else to control it eventually.

The var is a number, not a boolean. You add or subtract from the number to adjust a mob's possessability. The reason for this is so if two things make a mob possessable, the mob will stay possessable until both of them retract their possessability, not just any one of them.

When someone is possessable they look like this: 
possessable

This effect also applies to other things ghosts can click on:
ashwalker_eggs_interaction
clockwork_interaction
drone_interaction

### CODE REVIEW

File: code/modules/mob/living/living_defines.dm [hunk 1]
```diff
@@ -79,3 +79,5 @@
 	var/last_words	//used for database logging
 
 	var/list/obj/effect/proc_holder/abilities = list()
+
+	var/ondemand_possessable = 0 //a number, if more than 0, a ghost can shift-click a living mob to possess it.
```
Comment (JJRcop) [inline code/modules/mob/living/living_defines.dm]: >shift click
excuse me

File: code/modules/mob/living/living_defines.dm [hunk 2]
```diff
@@ -79,3 +79,5 @@
 	var/last_words	//used for database logging
 
 	var/list/obj/effect/proc_holder/abilities = list()
+
+	var/ondemand_possessable = 0 //a number, if more than 0, a ghost can ctrl-click a living mob to possess it.
```
Comment (lzimann) [inline code/modules/mob/living/living_defines.dm]: `FALSE`

File: code/modules/mob/dead/observer/observer.dm [hunk 3]
```diff
@@ -782,6 +782,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
+				to_chat(src, "<span class='warning'>[target] is no longer possessable!</span>")
+		else
+			to_chat(src, "<span class='warning'>[target] was possessed by another ghost!</span>")
+
+/mob/dead/observer/examinate(atom/A as mob|obj|turf in view())
+	. = ..()
+	if(isliving(A))
+		var/mob/living/L = A
+		var/L_possessable = L.ondemand_possessable
+		if(isnum(L_possessable) && L_possessable > 0)
```
Comment (lzimann) [inline code/modules/mob/dead/observer/observer.dm]: just `if(L.ondemand_possessable)` is enough here, no need to declare an extra var

File: code/modules/mob/dead/observer/observer.dm [hunk 4]
```diff
@@ -782,6 +782,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 	if(isobserver(user) && check_rights(R_SPAWN))
 		change_mob_type( /mob/living/carbon/human , null, null, TRUE) //always delmob, ghosts shouldn't be left lingering
 
+/mob/dead/observer/proc/ondemand_possess(mob/living/target)
+	var/possessable = target.ondemand_possessable
+	if(!isnum(possessable) || possessable <= 0)
```
Comment (lzimann) [inline code/modules/mob/dead/observer/observer.dm]: same here, you only need `if(target.on_demand_possessable)`

File: code/modules/mob/dead/observer/observer.dm [hunk 5]
```diff
@@ -782,6 +782,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
+	var/confirmation = tgalert(src, control_message, "[target] can be possessed", "Yes", "No", Timeout=FALSE)
+	if(confirmation == "Yes" && !QDELETED(target))
+		if(target.key == original_key)
+			possessable = target.ondemand_possessable
+			if(isnum(possessable) && possessable > 0)
+				to_chat(target, "<span class='ghostalert'>Your mob has been taken over by a ghost!</span>")
+				message_admins("[key_name_admin(src)] has taken control of ([key_name_admin(target)])")
+				target.ghostize(0)
+				target.key = key
+				target.ondemand_possessable = 0
```
Comment (lzimann) [inline code/modules/mob/dead/observer/observer.dm]: `FALSE`

File: code/modules/mob/dead/observer/observer.dm [hunk 6]
```diff
@@ -782,6 +782,35 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
+	var/possessable = target.ondemand_possessable
+	if(!isnum(possessable) || possessable <= 0)
+		return
+	var/original_key = target.key
+	var/control_message = control_mob_message(target)
+	var/confirmation = tgalert(src, control_message, "[target] can be possessed", "Yes", "No", Timeout=FALSE)
+	if(confirmation == "Yes" && !QDELETED(target))
+		if(target.key == original_key)
+			possessable = target.ondemand_possessable
+			if(isnum(possessable) && possessable > 0)
```
Comment (lzimann) [inline code/modules/mob/dead/observer/observer.dm]: just `if(target.ondemand_possessable)`

**Comment by lzimann** [comment]
the var not being a bool makes absolute no sense here imo

File: code/modules/mob/dead/observer/observer.dm (hunk 5 above)
Comment (JJRcop) [reply code/modules/mob/dead/observer/observer.dm]: not a boolea

File: code/modules/mob/dead/observer/observer.dm (hunk 4 above)
Comment (JJRcop) [reply code/modules/mob/dead/observer/observer.dm]: this isn't a boolea , it's a number. what if this is -1?

File: code/modules/mob/dead/observer/observer.dm (hunk 6 above)
Comment (JJRcop) [reply code/modules/mob/dead/observer/observer.dm]: it's a number, not boolean, what if it ends up being -1?

File: code/modules/mob/living/living_defines.dm (hunk 2 above)
Comment (JJRcop) [reply code/modules/mob/living/living_defines.dm]: not a boolean

**Comment by JJRcop** [comment]
the idea is, if two things make a mob possessable at the same time, by incrementing the var, then one of them removes that by subtracting one, the mob is still possessable until the other thing also subtracts one

**Comment by lzimann** [comment]
the subtraction never happens and i still don't think it makes any sense

**Comment by JJRcop** [comment]
this var isn't used yet, so subtraction would be done by things modifying the var. I will add this to antagbanned players when #31630 gets merged, and this will prevent collisions if another thing in the future makes you possessable temporarily then removes it, you're still possessable from the antagban

File: code/modules/mob/dead/observer/observer.dm (hunk 6 above)
Comment (Cyberboss) [reply code/modules/mob/dead/observer/observer.dm]: that's a bug then

Comment (JJRcop) [reply code/modules/mob/dead/observer/observer.dm]: I understand now, sorry

**Comment by duncathan** [comment]
i agree with the decision to not be a bool

**Comment by JJRcop** [comment]
replace_banned_player now uses possession if there is no volunteer ghost

**Comment by AnturK** [comment]
This should add some visual cue to ghosts or otherwise it's completely useless.

**Comment by JJRcop** [comment]
@AnturK K, I'll figure something out.

**Comment by JJRcop** [comment]
@AnturK 
possessable

**Comment by AnturK** [comment]
Neat, could you add it to existing mob spawners ? Would be consistent that way.

**Comment by JJRcop** [comment]
Sure

**Comment by JJRcop** [comment]
@AnturK
ashwalker_eggs_interaction
clockwork_interaction
drone_interaction

**Comment by AnturK** [comment]
I now realize this would be better as a componennt, but i guess it's bit late for this unless you feel like refactoring a bit.

**Comment by JJRcop** [comment]
I'm fine with making it a component, it's really not that big.

**Comment by RemieRichards** [comment]
Should be a bool.
Moronic to not be.

**Comment by JJRcop** [comment]
I componentized this, updating PR op.

File: code/datums/components/possessable.dm [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/datum/component/ondemand_possessable
+	var/amount = 0
+
+/datum/component/ondemand_possessable/Initialize()
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
```
Comment (ninjanomnom) [inline code/datums/components/possessable.dm]: This isn't how you component. Make the ghost clicking and displaying happen from inside here

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/_onclick/observer.dm [hunk 8]
```diff
@@ -48,6 +48,15 @@
 	// Not all of them require checking, see below
 	A.attack_ghost(src)
 
+/mob/dead/observer/CtrlClickOn(atom/A)
+	if(isliving(A))
+		var/mob/living/target = A
+		var/datum/component/ondemand_possessable/target_possessable = target.GetComponent(/datum/component/ondemand_possessable)
+		if(target_possessable)
+			ondemand_possess(target)
```
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/_onclick/observer.dm]: Oh wait the control click signal sends to the mob doing the clicking... Make a new signal for control click on the target.

[the remark this answers is not in the document]
Comment (JJRcop) [reply code/_onclick/observer.dm]: I'm turning this into a status effect

[the remark this answers is not in the document]
Comment (AnturK) [reply code/_onclick/observer.dm]: Yeah i meant something like AddComponent(../ghost_interaction,CALLBACK(src,./possess))
Where the component will pass the ghost user into the proc

**Comment by JJRcop** [comment]
I will componentize ghost interaction itself, which will involve removing attack_ghost and putting all the functionality into new procs for callback, but not in this PR. I've already violated the scope a little bit and I don't want to go any further with that right now.

**Comment by JJRcop** [comment]
I'm going to componentize this again but I'm going to do it right, it may take a while, because I have different priorities now

**Comment by optimumtact** [comment]
well feel free to make a new pr when you are ready, this one hasbeen stale too long

**Pull request closed without merging**